File: js/functions/helpers/goldCountChange.js
const goldCountWrapper = document.querySelector('.js-gold-count-wrapper'),
    goldCount = document.querySelector('.js-gold-count');

let goldIncreaseIterationCount = 0;

function goldCountChange(changeType) {
    const cheatBlock = document.querySelector('.js-cheats');
    if (!cheatBlock.classList.contains('active')) {
        let type = '';
        let number = 0;
        switch (changeType) {
            case 'increase':
                type = 'increase';
                number = 10;
                break;
            case 'decrease_first_letter':
                type = 'decrease';
                number = 3;
                break;
            case 'decrease_letter_count':
                type = 'decrease';
                number = 5;
                break;
            case 'decrease_max':
                type = 'decrease';
                number = 10;
                break;
            default:
                console.log('Неверный тип изменения');
        }

        if (goldCount.innerHTML - number <= -1 && type === 'decrease') {
            if (!goldCountWrapper.classList.contains('error')) {
                goldCountWrapper.classList.add('error');
                setTimeout(()=> goldCountWrapper.classList.remove('error'), 3000);
            }
            return 'not enough count'
        }

        else {
            goldCountWrapper.classList.add('active');

            setTimeout(() => goldCountWrapper.classList.remove('active'), 3000);

            changeIteration();
        }

        function changeIteration() {
            if (goldIncreaseIterationCount < number) {
                setTimeout(() => {
                    goldCount.innerHTML = changeType === 'increase' ? parseInt(goldCount.innerHTML) + 1 : goldCount.innerHTML - 1;
                    goldIncreaseIterationCount++;
                    changeIteration();
                }, 80);
            } else {
                goldIncreaseIterationCount = 0;
            }
        }
    }
}

export default goldCountChange;

File: cpp/goldCountChange.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

struct GoldCountData {
    int goldCount;
    int goldIncreaseIterationCount;
};

class GoldCounter {
public:
    GoldCountData data;

    GoldCounter() {
        data = {0, 0};
    }

    void changeGoldCount(const std::string& changeType) {
        if (changeType != "increase" && changeType != "decrease_first_letter" &&
            changeType != "decrease_letter_count" && changeType != "decrease_max") {
            std::cout << "Неверный тип изменения" << std::endl;
            return;
        }

        int number = 0;
        if (changeType == "increase") {
            number = 10;
        } else if (changeType == "decrease_first_letter") {
            number = 3;
        } else if (changeType == "decrease_letter_count") {
            number = 5;
        } else if (changeType == "decrease_max") {
            number = 10;
        }

        if (data.goldCount - number <= -1 && changeType.find("decrease") != std::string::npos) {
            std::cout << "not enough count" << std::endl;
            return;
        }

        for (int i = 0; i < number; ++i) {
            if (changeType == "increase") {
                data.goldCount++;
            } else {
                data.goldCount--;
            }
            data.goldIncreaseIterationCount++;
        }
    }
};